In [8]:
import pandas as pd

df = pd.read_json("/Users/kshitij/ML project/Dataset.json", lines=True) 
print(df.head(5))


          id           submitter  \
0  0704.0001      Pavel Nadolsky   
1  0704.0002        Louis Theran   
2  0704.0003         Hongjun Pan   
3  0704.0004        David Callan   
4  0704.0005  Alberto Torchinsky   

                                             authors  \
0  C. Bal\'azs, E. L. Berger, P. M. Nadolsky, C.-...   
1                    Ileana Streinu and Louis Theran   
2                                        Hongjun Pan   
3                                       David Callan   
4           Wael Abu-Shammala and Alberto Torchinsky   

                                               title  \
0  Calculation of prompt diphoton production cros...   
1           Sparsity-certifying Graph Decompositions   
2  The evolution of the Earth-Moon system based o...   
3  A determinant of Stirling cycle numbers counts...   
4  From dyadic $\Lambda_{\alpha}$ to $\Lambda_{\a...   

                                  comments  \
0  37 pages, 15 figures; published version   
1    To appear in

In [10]:
print(df.shape)
print(df.columns)

(2989022, 14)
Index(['id', 'submitter', 'authors', 'title', 'comments', 'journal-ref', 'doi',
       'report-no', 'categories', 'license', 'abstract', 'versions',
       'update_date', 'authors_parsed'],
      dtype='str')


In [11]:
print(df.shape)
print(df["abstract"][0])

(2989022, 14)
  A fully differential calculation in perturbative quantum chromodynamics is
presented for the production of massive photon pairs at hadron colliders. All
next-to-leading order perturbative contributions from quark-antiquark,
gluon-(anti)quark, and gluon-gluon subprocesses are included, as well as
all-orders resummation of initial-state gluon radiation valid at
next-to-next-to-leading logarithmic accuracy. The region of phase space is
specified in which the calculation is most reliable. Good agreement is
demonstrated with data from the Fermilab Tevatron, and predictions are made for
more detailed tests with CDF and DO data. Predictions are shown for
distributions of diphoton pairs produced at the energy of the Large Hadron
Collider (LHC). Distributions of the diphoton pairs from the decay of a Higgs
boson are contrasted with those produced from QCD processes at the LHC, showing
that enhanced sensitivity to the signal can be obtained with judicious
selection of events.



In [12]:
df_small = df.head(5000)
df_small = df_small[["abstract", "title", "categories"]]

print(df_small.shape)

(5000, 3)


In [15]:
df_small.isnull().sum()

abstract      0
title         0
categories    0
dtype: int64

In [16]:
df_small["abstract"][0]

'  A fully differential calculation in perturbative quantum chromodynamics is\npresented for the production of massive photon pairs at hadron colliders. All\nnext-to-leading order perturbative contributions from quark-antiquark,\ngluon-(anti)quark, and gluon-gluon subprocesses are included, as well as\nall-orders resummation of initial-state gluon radiation valid at\nnext-to-next-to-leading logarithmic accuracy. The region of phase space is\nspecified in which the calculation is most reliable. Good agreement is\ndemonstrated with data from the Fermilab Tevatron, and predictions are made for\nmore detailed tests with CDF and DO data. Predictions are shown for\ndistributions of diphoton pairs produced at the energy of the Large Hadron\nCollider (LHC). Distributions of the diphoton pairs from the decay of a Higgs\nboson are contrasted with those produced from QCD processes at the LHC, showing\nthat enhanced sensitivity to the signal can be obtained with judicious\nselection of events.\n'

In [18]:
df_small["abstract"] = df_small["abstract"].str.replace("\n", " ")
df_small["abstract"] = df_small["abstract"].str.strip()
df_small["abstract"][0]

'A fully differential calculation in perturbative quantum chromodynamics is presented for the production of massive photon pairs at hadron colliders. All next-to-leading order perturbative contributions from quark-antiquark, gluon-(anti)quark, and gluon-gluon subprocesses are included, as well as all-orders resummation of initial-state gluon radiation valid at next-to-next-to-leading logarithmic accuracy. The region of phase space is specified in which the calculation is most reliable. Good agreement is demonstrated with data from the Fermilab Tevatron, and predictions are made for more detailed tests with CDF and DO data. Predictions are shown for distributions of diphoton pairs produced at the energy of the Large Hadron Collider (LHC). Distributions of the diphoton pairs from the decay of a Higgs boson are contrasted with those produced from QCD processes at the LHC, showing that enhanced sensitivity to the signal can be obtained with judicious selection of events.'

In [20]:
df_small["abstract"] = df_small["abstract"].str.lower()
df_small["abstract"][0]

'a fully differential calculation in perturbative quantum chromodynamics is presented for the production of massive photon pairs at hadron colliders. all next-to-leading order perturbative contributions from quark-antiquark, gluon-(anti)quark, and gluon-gluon subprocesses are included, as well as all-orders resummation of initial-state gluon radiation valid at next-to-next-to-leading logarithmic accuracy. the region of phase space is specified in which the calculation is most reliable. good agreement is demonstrated with data from the fermilab tevatron, and predictions are made for more detailed tests with cdf and do data. predictions are shown for distributions of diphoton pairs produced at the energy of the large hadron collider (lhc). distributions of the diphoton pairs from the decay of a higgs boson are contrasted with those produced from qcd processes at the lhc, showing that enhanced sensitivity to the signal can be obtained with judicious selection of events.'

In [23]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(max_features=5000)

tfidf_matrix = vectorizer.fit_transform(df_small["abstract"])
print(tfidf_matrix.shape)

(5000, 5000)


In [26]:
from sklearn.metrics.pairwise import cosine_similarity

paper_index = 0;
print("Paper Title:", df_small["title"][paper_index])

scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()

print(scores.shape)

Paper Title: Calculation of prompt diphoton production cross sections at Tevatron and
  LHC energies
(5000,)


In [36]:
similar_indices = scores.argsort()

top_5 = similar_indices[-6:-1][::-1]

print("Top 5 similar papers to", df_small["title"][paper_index])
print

for i in top_5:                                                                                                                                              
    print(f"Score: {scores[i]:.4f} | {df_small['title'][i]}")   

Top 5 similar papers to Calculation of prompt diphoton production cross sections at Tevatron and
  LHC energies
Score: 0.3814 | Higgs boson production with one bottom quark including higher-order
  soft-gluon corrections
Score: 0.3762 | Resummation Effects in the Search of SM Higgs Boson at Hadron Colliders
Score: 0.3356 | Gluon Radiation of an Expanding Color Skyrmion in the Quark-Gluon Plasma
Score: 0.3322 | Resummed Cross Section for Jet Production at Hadron Colliders
Score: 0.3172 | Enhancement of prompt photons in ultrarelativistic proton-proton
  collisions from nonlinear gluon evolution at small-$x$


In [47]:
def recommend(paper_index, num_recommendations=5):
    scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()
    similar_indices = scores.argsort()[-6:-1][::-1]
    print("Recommendations for:", df_small["title"][paper_index])
    print(f"  Abstract: {df_small['abstract'][paper_index][:200]}")                                                                                                                                  
    print("---")                                                                                                                                             
    for i in similar_indices:
        print(f"Score: {scores[i]:.4f} | {df_small['title'][i]}")
        print(f"Abstract: {df_small['abstract'][i][:200]}")                                                                                                                                            


recommend(0)



Recommendations for: Calculation of prompt diphoton production cross sections at Tevatron and
  LHC energies
  Abstract: a fully differential calculation in perturbative quantum chromodynamics is presented for the production of massive photon pairs at hadron colliders. all next-to-leading order perturbative contribution
---
Score: 0.3814 | Higgs boson production with one bottom quark including higher-order
  soft-gluon corrections
Abstract: a higgs boson produced in association with one or more bottom quarks is of great theoretical and experimental interest to the high-energy community. a precise prediction of its total and differential 
Score: 0.3762 | Resummation Effects in the Search of SM Higgs Boson at Hadron Colliders
Abstract: we examine the soft-gluon resummation effects, including the exact spin correlations among the final state particles, in the search of the standard model higgs boson, via the process $gg\to h\to ww/zz
Score: 0.3356 | Gluon Radiation of an Expanding Color S

In [48]:
for n in [100, 500, 1000, 5000]:                                                                                                                             
      vec = TfidfVectorizer(max_features=n)                                                                                                                    
      matrix = vec.fit_transform(df_small["abstract"])                                                                                                         
      s = cosine_similarity(matrix[0], matrix).flatten()                                                                                                       
      top = s.argsort()[-2]  # best match (excluding itself)
      print(f"max_features={n:>5} | Top score: {s[top]:.4f}")   

max_features=  100 | Top score: 0.7787
max_features=  500 | Top score: 0.5385
max_features= 1000 | Top score: 0.3786
max_features= 5000 | Top score: 0.3814


Sharp observation! You noticed that 1000 and 5000 gave almost the same score (0.3786 vs 0.3814). That means words beyond 1000 aren't really adding much value.                                                                                                                                                    
                                                                                                                                                
  But here's the truth — there's no magic number. It depends on:                                                                                               
  - Your dataset size                                                                                                                                          
  - How varied the topics are                                                                                                                                  
  - What you consider a "good" match
                                    
  The real way to pick is experimentation — which you just did! And your data suggests somewhere around 1000–5000 is the sweet spot for this dataset (scores   
  barely change after 1000).                                                                                                                                   
                                                                                                                                                   
  For now, let's stick with 5000 and move forward. The bigger improvements won't come from tuning this number — they'll come from smarter techniques.          
                  
  ---                                                                                                                                                          
  Phase 5: NLP Improvements
                           
  Remember the "Hilbert" problem? TF-IDF treats every word independently. It can't understand that:
  - "deep learning" is one concept, not two separate words                                                                                                     
  - "car" and "automobile" mean the same thing                                                                                                                 
                                                                                                                                                               
  One simple improvement: stop words removal. These are useless common words like "the", "is", "we", "this".                                                   
                                                                                                                                                               
  TfidfVectorizer can do this automatically:                                                                                                                   
                                                                                                                                                               
  vectorizer = TfidfVectorizer(max_features=5000, stop_words="_____")

In [49]:
vectorizer = TfidfVectorizer(max_features=5000, stop_words="english")
tfidf_matrix = vectorizer.fit_transform(df_small["abstract"])

def recommend(paper_index, num_recommendations=5):
    scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()
    similar_indices = scores.argsort()[-6:-1][::-1]
    print("Recommendations for:", df_small["title"][paper_index])
    print(f"  Abstract: {df_small['abstract'][paper_index][:200]}")                                                                                                                                  
    print("---")                                                                                                                                             
    for i in similar_indices:
        print(f"Score: {scores[i]:.4f} | {df_small['title'][i]}")
        print(f"Abstract: {df_small['abstract'][i][:200]}")   

recommend(0)

Recommendations for: Calculation of prompt diphoton production cross sections at Tevatron and
  LHC energies
  Abstract: a fully differential calculation in perturbative quantum chromodynamics is presented for the production of massive photon pairs at hadron colliders. all next-to-leading order perturbative contribution
---
Score: 0.3228 | Resummation Effects in the Search of SM Higgs Boson at Hadron Colliders
Abstract: we examine the soft-gluon resummation effects, including the exact spin correlations among the final state particles, in the search of the standard model higgs boson, via the process $gg\to h\to ww/zz
Score: 0.3044 | Multiple Parton Scattering in Nuclei: Quark-quark Scattering
Abstract: modifications to quark and antiquark fragmentation functions due to quark-quark (antiquark) double scattering in nuclear medium are studied systematically up to order \cal{o}(\alpha_{s}^2)$ in deeply 
Score: 0.2932 | Gluon Radiation of an Expanding Color Skyrmion in the Quark-Gluon Plasm

Interesting! The order changed and scores actually went down slightly.

  Question for you: Why do you think removing stop words made scores lower instead of higher?                                                                  
   
  Think about it — if both papers had the word "the" 20 times, that was adding to their similarity score. Now that's removed, so only the meaningful words     
  count.                                                                                                                                                     
                                                                                                                                                               
  But is a lower score with better matching actually a bad thing?                                                                                              
  
  Look at the new #2 result — "Multiple Parton Scattering in Nuclei" — that's a new paper that appeared, and it's actually about quark scattering in particle  
  physics, which is relevant!
                                                                                                                                                               
  The scores are lower but potentially more honest. Before, common words were inflating the scores.                                                            
  
  ---                                                                                                                                                          
  Let's try one more improvement — n-grams
                                          
  Right now TF-IDF looks at single words. But "machine learning" as a pair is more meaningful than "machine" and "learning" separately.
                                                                                                                                                               
  vectorizer = TfidfVectorizer(max_features=5000, stop_words="english", ngram_range=(1, ___))

In [50]:
vectorizer = TfidfVectorizer(max_features=5000, stop_words="english", ngram_range=(1, 2))
tfidf_matrix = vectorizer.fit_transform(df_small["abstract"])

def recommend(paper_index, num_recommendations=5):
    scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()
    similar_indices = scores.argsort()[-6:-1][::-1]
    print("Recommendations for:", df_small["title"][paper_index])
    print(f"  Abstract: {df_small['abstract'][paper_index][:200]}")                                                                                                                                  
    print("---")                                                                                                                                             
    for i in similar_indices:
        print(f"Score: {scores[i]:.4f} | {df_small['title'][i]}")
        print(f"Abstract: {df_small['abstract'][i][:200]}")   

recommend(0)

Recommendations for: Calculation of prompt diphoton production cross sections at Tevatron and
  LHC energies
  Abstract: a fully differential calculation in perturbative quantum chromodynamics is presented for the production of massive photon pairs at hadron colliders. all next-to-leading order perturbative contribution
---
Score: 0.3345 | Resummation Effects in the Search of SM Higgs Boson at Hadron Colliders
Abstract: we examine the soft-gluon resummation effects, including the exact spin correlations among the final state particles, in the search of the standard model higgs boson, via the process $gg\to h\to ww/zz
Score: 0.3233 | Multiple Parton Scattering in Nuclei: Quark-quark Scattering
Abstract: modifications to quark and antiquark fragmentation functions due to quark-quark (antiquark) double scattering in nuclear medium are studied systematically up to order \cal{o}(\alpha_{s}^2)$ in deeply 
Score: 0.3126 | Higgs boson production with one bottom quark including higher-order
  s

 ---                                                                                                                                                          
  Phase 6: Evaluation
                                                                                                                                                               
  Right now we just "look" at results and guess if they're good. A real ML system needs a metric — a number that tells you how good your system is.
                                                                                                                                                               
  Question for you: We have a categories column in our data. Each paper has a category like "hep-ph" (particle physics), "math.CO" (math), etc.                
                                                                                                                                                               
  How could we use this column to measure if our recommendations are actually good?                                                                            
                  
  Think simply — if I recommend 5 papers for a physics paper, what would be a sign that the recommendations are good?                                          
                  
❯ i guess we will match how many categories word match  

  1. Pick a paper → get its categories                                                                                                                         
  2. Find 5 recommended papers                                                                                                                                 
  3. Check each recommendation — does it share any category with the source?                                                                                   
  4. Count how many matched → divide by 5                                                                                                                      
                                                                                                                                                               
  Then we do this for 100 papers and average the score to get overall accuracy.    

In [58]:
def evaluate(paper_index):                                                                                                                                   
    scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()                                                                            
    similar_indices = scores.argsort()[-6:-1][::-1]                                                                                                          
                                                                                                                                                               
    source_cats = set(df_small["categories"][paper_index].split(" "))                                                                                              
    matches = 0                                                                                                                                              
                                                                                                                                                               
    for i in similar_indices:                       
        rec_cats = set(df_small["categories"][i].split(" "))                                                                                                         
        if source_cats.intersection(rec_cats):  
            matches += 1
   
    return matches / 5  # accuracy out of 1.0                                                                                                                
                  
  # Test on first 100 papers and average the score                                                                                                             
total = 0       
for i in range(100):                                                                                                                                         
    total += evaluate(i)

print(f"Average accuracy: {total/100:.2f}")   

Average accuracy: 0.39


fit_transform does 2 things:                                                                                                                                 
  1. fit — learn the vocabulary (which words exist, their importance)                                                                                          
  2. transform — convert text into numbers using that vocabulary                                                                                               
                                                                                                                                                               
  transform does only step 2 — convert text using the already learned vocabulary.                                                                              
                                                                                                                                                               
  Why does this matter?                   
                                                                                                                                                               
  #### When building the system — learn + convert                                                                                                                 
  tfidf_matrix = vectorizer.fit_transform(df_small["abstract"])                                                                                                
  ####"I've never seen this data. Let me learn the words, then convert."                                                                                         
                                                            
  #### When a user searches — just convert                                                                                                                        
  query_vector = vectorizer.transform(["deep learning"])                                                                                                       
  #### "I already know the words. Just convert this new text."                                                                                                    
                                                                                                                                                               
  If you used fit_transform on the query, it would forget everything it learned from the 5000 papers and learn only from your tiny query. That would break the 
  comparison.                                                                                                                                                  
   
  ---                                                                                                                                                          
  Now fill in both blanks and run it. Test with:            
  search("deep learning")   

In [60]:
def search(query, num_results=5):                                                                                                                            
      query_vector = vectorizer.transform([query])                                                                                                                 
      scores = cosine_similarity(query_vector, tfidf_matrix).flatten()                                                                                                
                                                                                                                                                               
      top_indices = scores.argsort()[-num_results:][::-1]   

      for i in top_indices:                                                                                                                                    
          print(f"Score: {scores[i]:.4f} | {df_small['title'][i]}")
          print(f"  Abstract: {df_small['abstract'][i][:150]}")                                                                                                
          print()  

search("deep learning for natural language processing")

Score: 0.3093 | Using conceptual metaphor and functional grammar to explore how language
  used in physics affects student learning
  Abstract: this paper introduces a theory about the role of language in learning physics. the theory is developed in the context of physics students' and physici

Score: 0.3081 | Birth, survival and death of languages by Monte Carlo simulation
  Abstract: simulations of physicists for the competition between adult languages since 2003 are reviewed. how many languages are spoken by how many people? how m

Score: 0.3077 | An Adaptive Strategy for the Classification of G-Protein Coupled
  Receptors
  Abstract: one of the major problems in computational biology is the inability of existing classification models to incorporate expanding and new domain knowledg

Score: 0.2954 | Ensemble Learning for Free with Evolutionary Algorithms ?
  Abstract: evolutionary learning proceeds by evolving a population of classifiers, from which it generally returns (with some n

#### The result is not good
but How many abstracts even mention "deep learning"?   


In [65]:
print("deep learning:", df_small["abstract"].str.contains("deep learning").sum())                                                                            
print("neural network:", df_small["abstract"].str.contains("neural network").sum())   

deep learning: 0
neural network: 16


  Let's do Phase 7 now: Scale up               (*it was done after phase 8)                                                                                                             
  Instead of taking the first 5000 papers (all from 2007), let's take a random sample that covers all years and topics

In [80]:
# Reload the full dataset               
df = pd.read_json("/Users/kshitij/ML project/Dataset.json", lines=True)
                                                                                                                                                               
# Take a random sample of 100,000 papers
df_large = df.sample(n=100000, random_state=42)   
# As indices will be jumbled, reset them to be sequential  
df_large = df_large.reset_index(drop=True)                                                                                                          
                                                                                                                                                               
# Check the spread of years                 
print(df_large["update_date"].min(), "to", df_large["update_date"].max())                                                                                    
                                                            
# Check for deep learning now                                                                                                                                
print("deep learning:", df_large["abstract"].str.contains("deep learning").sum())

Exception ignored in: <bound method IPythonKernel._clean_thread_parent_frames of <ipykernel.ipkernel.IPythonKernel object at 0x107be1be0>>
Traceback (most recent call last):
  File "/Users/kshitij/LLMOPS/.venv/lib/python3.13/site-packages/ipykernel/ipkernel.py", line 781, in _clean_thread_parent_frames
    def _clean_thread_parent_frames(
KeyboardInterrupt: 


2007-05-23 to 2026-03-20
deep learning: 1485


In [81]:
df_large = df_large[["title", "abstract", "categories"]]

  # Clean abstracts — you've done all these before
df_large["abstract"] = df_large["abstract"].str.replace("\n", " ")
df_large["abstract"] = df_large["abstract"].str.strip()
df_large["abstract"] = df_large["abstract"].str.lower()


  # Check for missing values
print(df_large.isnull().sum())

title         0
abstract      0
categories    0
dtype: int64


In [82]:
vectorizer = TfidfVectorizer(max_features=5000, stop_words="english", ngram_range=(1, 2))
tfidf_matrix = vectorizer.fit_transform(df_large["abstract"])

print(tfidf_matrix.shape)

(100000, 5000)


In [85]:
def search(query, num_results=5):                                                                                                                            
      query_vector = vectorizer.transform([query])                                                                                                                 
      scores = cosine_similarity(query_vector, tfidf_matrix).flatten()                                                                                                
                                                                                                                                                               
      top_indices = scores.argsort()[-num_results:][::-1]   

      for i in top_indices:                                                                                                                                    
          print(f"Score: {scores[i]:.4f} | {df_large['title'][i]}")
          print(f"  Abstract: {df_large['abstract'][i][:150]}")                                                                                                
          print()

search("deep learning for natural language processing")
print("\n\n")
search("quantum physics")
print("\n\n")
search("image recognition computer vision")

Score: 0.4301 | Turkish Sentiment Analysis Using Machine Learning Methods: Application
  on Online Food Order Site Reviews
  Abstract: satisfaction measurement, which emerges in every sector today, is a very important factor for many companies. in this study, it is aimed to reach the 

Score: 0.4131 | Deep Learning Works in Practice. But Does it Work in Theory?
  Abstract: deep learning relies on a very specific kind of neural networks: those superposing several neural layers. in the last few years, deep learning achieve

Score: 0.4094 | NLML--a Markup Language to Describe the Unlimited English Grammar
  Abstract: in this paper we present nlml (natural language markup language), a markup language to describe the syntactic and semantic structure of any grammatica

Score: 0.4091 | Structure preserving deep learning
  Abstract: over the past few years, deep learning has risen to the foreground as a topic of massive interest, mainly as a result of successes obtained in solving

Score: 0.40

In [90]:
def evaluate(paper_index):
    scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()
    similar_indices = scores.argsort()[-6:-1][::-1]

    source_cats = set(df_large["categories"][paper_index].split(" "))
    matches = 0

    for i in similar_indices:
        rec_cats = set(df_large["categories"][i].split(" "))
        if source_cats.intersection(rec_cats):
            matches += 1

    return matches / 5

    total = 0
    for i in range(500):
        total += evaluate(i)

print(f"Average accuracy: {total/500:.2f}")

Average accuracy: 0.08


In [95]:
l_all_cats = set()
for cats in df_large["categories"]:
    for c in cats.split(" "):
        l_all_cats.add(c)
        
s_all_cats = set()
for cats in df_small["categories"]:
    for c in cats.split(" "):
        s_all_cats.add(c)
        
print("Unique individual categories_df_Large:", len(l_all_cats))
print("Unique individual categories_df_Small:", len(s_all_cats))

Unique individual categories_df_Large: 175
Unique individual categories_df_Small: 135


 175 vs 135 — that's not a huge difference. So the problem isn't the number of categories.

  Let's dig deeper. Run the diagnostic code to see what's actually happening:

In [ ]:
i = 0
scores = cosine_similarity(tfidf_matrix[i], tfidf_matrix).flatten()
similar_indices = scores.argsort()[-6:-1][::-1]
print("Source:", df_large["categories"][i])

print("---")
for idx in similar_indices:
    print(f"Rec: {df_large['categories'][idx]} | Score: {scores[idx]:.4f}")

Source: cs.GR
---
Rec: physics.comp-ph cond-mat.soft eess.IV | Score: 0.4212
Rec: cond-mat.soft | Score: 0.3643
Rec: physics.ins-det | Score: 0.3416
Rec: physics.flu-dyn cond-mat.soft | Score: 0.3401
Rec: cs.CR | Score: 0.3349


Source is cs.GR (Computer Graphics) but recommendations are physics.comp-ph, cond-mat.soft, physics.flu-dyn — none share the category.

  But wait — are these actually bad recommendations? Let's check what the papers are actually about:

In [99]:
print("Source title:", df_large["title"][i])
print("Source abstract:", df_large["abstract"][i][:200])
print("Source categories:", df_large["categories"][i])
print("---")
for idx in similar_indices:
    print(f"Category: {df_large['categories'][idx]}")
    print(f"Title: {df_large['title'][idx]}")
    print(f"Abstract: {df_large['abstract'][idx][:200]}")

Source title: NeuralDrop: DNN-based Simulation of Small-Scale Liquid Flows on Solids
Source abstract: small-scale liquid flows on solid surfaces provide convincing details in liquid animation, but they are difficult to be simulated with efficiency and fidelity, mostly due to the complex nature of the 
Source categories: cs.GR
---
Category: physics.comp-ph cond-mat.soft eess.IV
Title: Learning physical properties of liquid crystals with deep convolutional
  neural networks
Abstract: machine learning algorithms have been available since the 1990s, but it is much more recently that they have come into use also in the physical sciences. while these algorithms have already proven to 
Category: cond-mat.soft
Title: Novel method for determination of contact angle of highly volatile
  liquids
Abstract: in this paper, a novel method for the measurement of equilibrium contact angle of highly volatile binary liquids is proposed. the proposed method, which combines finite element method and energy

Instead of matching cs.GR with physics.flu-dyn, we match the domain prefix:

  - cs.GR → domain is cs
  - physics.flu-dyn → domain is physics
  - cond-mat.soft → domain is cond-mat

  But even this won't help for the liquid example — cs vs physics are different domains even though the papers are related.

  A better approach: check if the top words overlap between source and recommendations. But that's basically what TF-IDF already does!

  Here's a practical middle ground — let's try both metrics and compare:

In [107]:
def evaluate_v2(paper_index):
    scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()
    similar_indices = scores.argsort()[-6:-1][::-1]

    source_cats = set(df_large["categories"][paper_index].split(" "))
    source_domains = set(c.split(".")[0] for c in source_cats)

    cat_matches = 0
    domain_matches = 0

    for i in similar_indices:
        rec_cats = set(df_large["categories"][i].split(" "))
        rec_domains = set(c.split(".")[0] for c in rec_cats)

    if source_cats.intersection(rec_cats):
        cat_matches += 1
    if source_domains.intersection(rec_domains):
            domain_matches += 1

    return cat_matches / 5, domain_matches / 5

# Test on 500 papers
cat_total, domain_total = 0, 0
for i in range(500):
    c, d = evaluate_v2(i)
    cat_total += c
    domain_total += d
print(f"Category accuracy: {cat_total/500:.2f}")
print(f"Domain accuracy: {domain_total/500:.2f}")

Category accuracy: 0.10
Domain accuracy: 0.15


For the resume, we can say:
"Achieved 15% domain-level accuracy and 10% category-level accuracy, with analysis showing cross-disciplinary recommendations that are topically relevant
  but span different category labels"

CountVectorizer does NOT measure rarity. That's the difference.

  - CountVectorizer — just counts how many times each word appears. "the" appearing 50 times gets a score of 50.
  - TfidfVectorizer — counts words AND penalizes common ones. "the" appearing everywhere gets a low score despite being frequent.

In [111]:
from sklearn.feature_extraction.text import CountVectorizer

count_vec = CountVectorizer(max_features=5000, stop_words="english", ngram_range=(1, 2))
count_matrix = count_vec.fit_transform(df_large["abstract"])

def evaluate_count(paper_index):
    scores = cosine_similarity(count_matrix[paper_index], count_matrix).flatten()
    similar_indices = scores.argsort()[-6:-1][::-1]

    source_cats = set(df_large["categories"][paper_index].split(" "))
    source_domains = set(c.split(".")[0] for c in source_cats)

    cat_matches = 0
    domain_matches = 0

    for i in similar_indices:
        rec_cats = set(df_large["categories"][i].split(" "))
        rec_domains = set(c.split(".")[0] for c in rec_cats)

        if source_cats.intersection(rec_cats):
            cat_matches += 1
        if source_domains.intersection(rec_domains):
            domain_matches += 1

    return cat_matches / 5, domain_matches / 5

cat_total, domain_total = 0, 0
for i in range(500):
    c, d = evaluate_count(i)
    cat_total += c
    domain_total += d

print(f"CountVec - Category accuracy: {cat_total/500:.2f}")
print(f"CountVec - Domain accuracy: {domain_total/500:.2f}")

CountVec - Category accuracy: 0.49
CountVec - Domain accuracy: 0.77


 CountVectorizer crushed TF-IDF! Your prediction was right!

  But wait — why? This is surprising. TF-IDF is supposed to be smarter. Let me ask you:

  Think about it: CountVectorizer gives high scores to common words within a domain. So papers in physics all use words like "particle", "energy", "field" a
  lot — and CountVectorizer scores those highly. That makes papers in the same domain match strongly.

  TF-IDF penalizes those common domain words because they appear across many papers — which actually hurts same-domain matching.

  ---
  The lesson:

  "Smarter" isn't always better. It depends on what you're measuring.
  - TF-IDF finds papers about the same specific topic (cross-disciplinary)
  - CountVectorizer finds papers in the same domain (same category)

 let's use CountVectorizer as our main approach since it performs better on our metric.

In [113]:
def search(query, num_results=5):
    query_vector = count_vec.transform([query])
    scores = cosine_similarity(query_vector, count_matrix).flatten()

    top_indices = scores.argsort()[-num_results:][::-1]

    for i in top_indices:
        print(f"Score: {scores[i]:.4f} | {df_large['title'][i]}")
        print(f"  Category: {df_large['categories'][i]}")
        print()

search("deep learning for natural language processing")

Score: 0.4983 | Performance Evaluation of Deep Learning Tools in Docker Containers
  Category: cs.DC cs.LG cs.PF

Score: 0.4838 | Structure preserving deep learning
  Category: cs.LG cs.NA math.NA stat.ML

Score: 0.4634 | Deep Learning Works in Practice. But Does it Work in Theory?
  Category: cs.AI

Score: 0.4575 | Using NLU in Context for Question Answering: Improving on Facebook's
  bAbI Tasks
  Category: cs.CL cs.AI

Score: 0.4456 | Long-range and hierarchical language predictions in brains and
  algorithms
  Category: q-bio.NC cs.AI cs.CL cs.LG cs.NE



## Compared TF-IDF and CountVectorizer approaches — CountVectorizer achieved 77% domain accuracy vs TF-IDF's 50%, demonstrating that word frequency without IDF weighting better captures domain-level similarity

mixed_scores = (tfidf_scores + count_scores) / 2

  This combines both approaches:
  - count_scores pulls papers from the same domain (its strength)
  - tfidf_scores pulls papers about the same specific topic (its strength)
  - Averaging them means a paper needs to score well on both to rank high

  The result: recommendations that are in the right domain AND about the right topic.

  ---
  A more advanced version would be weighted — giving one approach more influence:

  ⏺ Look at this! Full comparison: for 50% CountVec (domain) + 50% TF-IDF (topic)

   │    Approach     │ Category │ Domain │
  ├─────────────────┼──────────┼────────┤
  │ TF-IDF          │ 0.10     │ 0.15   │
  ├─────────────────┼──────────┼────────┤
  │ CountVectorizer │ 0.49     │ 0.77   │
  ├─────────────────┼──────────┼────────┤
  │ Mixed           │ 0.50     │ 0.78   │
  └─────────────────┴──────────┴────────┘

  Mixed is slightly better than CountVectorizer alone on both metrics! The TF-IDF component is adding a small boost by catching topic-specific matches.

  #### 70% CountVec (domain) + 30% TF-IDF (topic)
  mixed_scores = (0.7 * count_scores) + (0.3 * tfidf_scores)

   Almost the same — 0.49/0.78 vs 0.50/0.78. The 50/50 split was actually slightly better!

  This tells us the equal average is the sweet spot for this dataset. No need to overcomplicate it.

  ---
  Final scorecard for your resume:

  ┌─────────────────┬──────────┬────────┐
  │    Approach     │ Category │ Domain │
  ├─────────────────┼──────────┼────────┤
  │ TF-IDF          │ 0.10     │ 0.15   │
  ├─────────────────┼──────────┼────────┤
  │ CountVectorizer │ 0.49     │ 0.77   │
  ├─────────────────┼──────────┼────────┤
  │ Mixed (best)    │ 0.50     │ 0.78   │
  └─────────────────┴──────────┴────────┘

  This is a great story: "Tested 3 approaches, found that combining TF-IDF and CountVectorizer achieved the best results — 78% domain accuracy."

In [117]:
def evaluate_mixed(paper_index):
    tfidf_scores = cosine_similarity(tfidf_matrix[paper_index], tfidf_matrix).flatten()
    count_scores = cosine_similarity(count_matrix[paper_index], count_matrix).flatten()

    # Combine both scores (average)
    mixed_scores = 0.3*tfidf_scores + 0.7*count_scores

    similar_indices = mixed_scores.argsort()[-6:-1][::-1]

    source_cats = set(df_large["categories"][paper_index].split(" "))
    source_domains = set(c.split(".")[0] for c in source_cats)

    cat_matches = 0
    domain_matches = 0

    for i in similar_indices:
        rec_cats = set(df_large["categories"][i].split(" "))
        rec_domains = set(c.split(".")[0] for c in rec_cats)

        if source_cats.intersection(rec_cats):
            cat_matches += 1
        if source_domains.intersection(rec_domains):
            domain_matches += 1

    return cat_matches / 5, domain_matches / 5

cat_total, domain_total = 0, 0
for i in range(500):
    c, d = evaluate_mixed(i)
    cat_total += c
    domain_total += d

print(f"Mixed - Category accuracy: {cat_total/500:.2f}")
print(f"Mixed - Domain accuracy: {domain_total/500:.2f}")

Mixed - Category accuracy: 0.49
Mixed - Domain accuracy: 0.78
